README.md
# Vue Clazy Load

> _Claziest_ lazy loader out there!

Component-based image lazy loader for Vue.js 2

[![npm](https://img.shields.io/npm/dm/vue-clazy-load.svg)](https://www.npmjs.com/package/vue-clazy-load)
[![npm](https://img.shields.io/npm/v/vue-clazy-load.svg)](https://www.npmjs.com/package/vue-clazy-load)
[![npm](https://img.shields.io/npm/l/vue-clazy-load.svg)](https://www.npmjs.com/package/vue-clazy-load)
[![GitHub pull requests](https://img.shields.io/badge/PR-welcome-green.svg)]()
[![GitHub stars](https://img.shields.io/github/stars/matheusgrieger/vue-clazy-load.svg?style=social&label=Star)]()

Swaps between your image and another component when loading images, allowing you to use loaders from [component frameworks](https://github.com/vuejs/awesome-vue#frameworks) such as spinners and progress bars. This method also allows transitioning between the two components.

## Demo

Check out the [example page](https://matheusgrieger.github.io/vue-clazy-load/example/index.html).

## Installation

Install with npm or yarn:

```sh
npm install vue-clazy-load
yarn add vue-clazy-load
```

Then simply import it to your project through the method that suits you best

* ES6+
  ```js
  import VueClazyLoad from 'vue-clazy-load'
  ```

* Common/Require
  ```js
  var VueClazyLoad = require('vue-clazy-load')
  ```

And install into your Vue instance

```js
Vue.use(VueClazyLoad)
```

You can import it into specific components if you don't want to register Clazy Load globally

```js
import { VueClazyLoad } from 'vue-clazy-load'

export default {
  components: {
    VueClazyLoad
  }
}
```

Also available through Unpkg CDN

```html
<script src="https://unpkg.com/vue-clazy-load/dist/vue-clazy-load.min.js"></script>
```

## Documentation

Clazy Load works without any JS configuration as is, all you need is the basic HTML markup:

```html
<clazy-load src="https://unsplash.it/500">
  <img src="https://unsplash.it/500">
  <div class="preloader" slot="placeholder">
    Loading message
  </div>
</clazy-load>
```
The only required prop you must set is `src` that must correspond to your image's.

### Props

All props supported by Clazy Load are listed below with their types and explanation.

#### `src`

* Type: string
* Default: none
* _Required_

Source of the image to be loaded. Must match your `<img>` tag src.

#### `tag`

* Type: string
* Default: `"div"`

What tag the component should render to.

#### `element`

* Type: string
* Default: `null`

Selector for Intersecion Observer's root element. Leave blank/null to use _viewport_.

#### `threshold`

* Type: Array | number
* Default: `[0, 0.5, 1]`

Values for Intersection Observer's threshold option.

#### `ratio`

* Type: number
* Default: `0.4`

Percent of the element that needs to be visible to trigger loading. Must be > 0 and <= 1.

#### `margin`

* Type: string
* Default: `"0px"`

Intersection Observer's margin option.

You can read more on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) about every Intersection Observer-specific prop and what they do.

#### `crossorigin`

* Type: string
* Default: `null`
* Valid options: `"anonymous"` | `"use-credentials"`

Sets image's `crossOrigin` option and allows loading external images. Useful for Service Workers and caching.

#### `loadedClass`

* Type: string
* Default: `"loaded"`

#### `loadingClass`

* Type: string
* Default: `"loading"`

#### `errorClass`

* Type: string
* Default: `null`

All classes are added to the **root** element, not the image itself.

### Events

#### `loading`

Image started loading and placeholder is visible.

#### `load`

* Param: native load event

Image finished loading and is now visible.

#### `error`

* Param: native error event

Could not load image. **Image is not shown, placeholder still visible.**

### Compatibility

Vue Clazy Load uses the Intersection Observer API to watch for the element visibility on screen. The advantages are native optimization from each browser and no need to implement a custom solution that may be buggy and increase file size. The only caveat to this approach is that this API is quite new, so older browsers do not support it.

If your application needs to be backwards compatible with IE and others, there are polyfills available. I personally recommend [Polyfill.io](https://polyfill.io/). You can check their documentation on how to add it to your website, or simply include the following tag if you're not using any other polyfills:

```html
<script src="https://cdn.polyfill.io/v2/polyfill.min.js?features=IntersectionObserver"></script>
```

## Changelog

Check out [changelog](CHANGELOG.md) file.

## Roadmap

Check [roadmap](ROADMAP.md) file.

## Contributing

Issues, questions and feature requests are welcome. If you can cover some problem, pull requests are also very welcome!

## License

[MIT](http://opensource.org/licenses/MIT)

Copyright (c) 2017-2018, Matheus Grieger


CHANGELOG.md
# Changelog

## v0.4.2

- **FIX** restored install function for ES6 modules

## v0.4.1

- **FIX** accidentaly removed install function

## v0.4.0

- **NEW** custom loading, loaded and error class naming
- **NEW** support for CORS (thanks to @gerardo-rodriguez)
- **FIX** last update was not present on NPM, somehow
- **FIX** improved docs

## v0.3.0

- **NEW** extra `component` export so you can install the component where needed, not only globally
- **NEW** don't need to explicitly add `slot="image"` to your tags, the `default` slot will be used prioritarily, if found
  Note: if both are present, only the `default` will be used.
- **FIX** changed `requestAnimationFrame` to `$nextTick`, which reduces errors of undefined elements
- **FIX** removed main element `ref` to use `this.$el` instead, reducing processing

## v0.2.0

- **NEW** error event for when image fails to load
- **NEW** margin option for IntersectionObserver's `rootMargin`
- **NEW** ratio option for comparing element's visible percentage
- **FIX** (some) bad documentation grammar

## v0.1.5

- **FIX** build error when using Webpack and `import`ing

## v0.1.4

- **FIX** `import` was not working in Babel

## v0.1.3

- **FIX** double loading on fast scrolling and/or slow connections

## v0.1.2

- **FIX** Observer being `null`ified before disconnecting

## v0.1.1

Failed version.

## v0.1.0

Initial release.


src/clazy-load.js
/*!
 * Vue Clazy Load
 * Component-based lazy (CLazy) load images in Vue.js 2
 * @author Matheus Grieger
 * @version 0.4.2
 */
const ClazyLoadComponent = {
  name: 'ClazyLoad',
  props: {
    /**
     * HTML/Component tag name to be used in place of the component
     * @type {String}
     * @default div
     */
    tag: {
      type: String,
      default: 'div'
    },
    /**
     * Image source URL
     * @type {String}
     * @required
     */
    src: {
      type: String,
      required: true
    },
    /**
     * IntersectionObserver root element
     * @type {String}
     */
    element: String,
    /**
     * IntersectionObserver threshold
     * @type {Array, Number}
     */
    threshold: {
      type: [Array, Number],
      default: () => {
        return [0, 0.5, 1]
      }
    },
    /**
     * InserectionObserver visibility ratio
     * @type {Number}
     */
    ratio: {
      type: Number,
      default: 0.4,
      validator(value) {
        // can't be less or equal to 0 and greater than 1
        return value > 0 && value <= 1
      }
    },
    /**
     * IntersectionObserver root margin
     * @type {String}
     */
    margin: {
      type: String,
      default: '0px'
    },
    /**
     * Optional CORS mode ("anonymous" | "use-credentials")
     * @see https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_enabled_image
     * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-crossorigin
     * @type {String}
     */
    crossorigin: {
      type: String,
      default: null,
      validator(value) {
        return value === 'anonymous' || value === 'use-credentials'
      }
    },
    /**
     * Class added to element when it finishes loading
     * @type {String}
     * @default loaded
     */
    loadedClass: {
      type: String,
      default: 'loaded'
    },
    /**
     * Class added to element while it is loading
     * @type {String}
     */
    loadingClass: {
      type: String,
      default: 'loading'
    },
    /**
     * Class added to element if loading failed
     * @type {String}
     */
    errorClass: {
      type: String,
      default: null
    }
  },
  data() {
    return {
      loaded: false,
      observer: null,
      errored: false
    }
  },
  methods: {
    /**
     * Start loading image
     */
    load() {
      // emits 'loading' event upwards
      this.$emit('loading')

      // disconnect observer
      // so it doesn't load more than once
      this.observer.disconnect()

      if (!this.loaded) {
        // fake image
        let img = new Image

        img.addEventListener('load', () => {
          this.loaded = true
          // emits 'load' event upwards
          this.$emit('load')

          clear()
        })

        img.addEventListener('error', (event) => {
          this.errored = true
          // emits 'error' event upwards
          // adds the original event as argument
          this.$emit('error', event)

          clear()
        })

        // function used to clear variables from memory
        const clear = () => {
          // discard fake image
          img = null
          // remove observer from memory
          this.observer = null
        }

        // CORS mode configuration
        if (this.crossorigin !== null) {
          img.crossOrigin = this.crossorigin
        }

        img.src = this.src
      }
    },

    /**
     * Creates IntersectionObserver instance and observe current element
     */
    observe() {
      const options = {
        threshold: this.threshold,
        root: this.element ? document.querySelector(this.element) : null,
        rootMargin: this.margin
      }

      // creates IO instance
      this.observer = new IntersectionObserver((entries) => {
        // as we instantiated one for each component
        // we can directly access the first index
        if (entries[0].intersectionRatio >= this.ratio) {
          this.load()
        }
      }, options)

      // start observing main component
      this.observer.observe(this.$el)
    }
  },
  render(h) {
    // class to be added to element indicating load state
    const elementClass = this.loaded ? this.loadedClass : this.loadingClass

    return h(this.tag, {
      // if loading failed adds error class if exists,
      // otherwhise adds elementClass defined above
      class: this.errored && this.errorClass ? this.errorClass : elementClass
    }, [
      this.loaded
        ? this.$slots.default || this.$slots.image // allows for "default" slot
        : this.$slots.placeholder
    ])
  },
  mounted() {
    // start observing the element visibility
    this.$nextTick(this.observe)
  }
}

// Export install function for CDN embeds
export const install = (Vue) => {
  Vue.component('clazy-load', ClazyLoadComponent)
}

// Component object
export const VueClazyLoad = ClazyLoadComponent

// Exports default object for ES6 modules
export default {
  install
}
